src/optimizer.cpp

```diff
@@ -249,71 +250,99 @@ void Optimizer::propagateStateVelocitiesFromInitials(
 {
   motion_model_->predict(state);
 }
+
 void Optimizer::integrateStateVelocities(
-  xt::xtensor<float, 3> & trajectories,
-  const models::State & state) const
+  xt::xtensor<float, 2> & trajectories,
+  const xt::xtensor<float, 2> & state) const
 {
-  auto x = xt::view(trajectories, xt::all(), xt::all(), 0);
-  auto y = xt::view(trajectories, xt::all(), xt::all(), 1);
-  auto yaw = xt::view(trajectories, xt::all(), xt::all(), 2);
-
-  auto & w = state.wz;
   double initial_yaw = tf2::getYaw(state_.pose.pose.orientation);
 
-  yaw = utils::normalize_angles(xt::cumsum(w * settings_.model_dt, 1) + initial_yaw);
-  xt::xtensor<float, 2> yaw_offseted = yaw;
+  auto vx = xt::view(state, xt::all(), 0);
+  auto vy = xt::view(state, xt::all(), 2);
+  auto wz = xt::view(state, xt::all(), 1);
+
+  auto traj_x = xt::view(trajectories, xt::all(), 0);
+  auto traj_y = xt::view(trajectories, xt::all(), 1);
+  auto traj_yaws = xt::view(trajectories, xt::all(), 2);
 
-  xt::view(yaw_offseted, xt::all(), xt::range(1, _)) =
-    xt::view(yaw, xt::all(), xt::range(_, -1));
-  xt::view(yaw_offseted, xt::all(), 0) = initial_yaw;
+  traj_yaws = xt::cumsum(utils::normalize_angles(wz) * settings_.model_dt, 0) + initial_yaw;
 
-  xt::xtensor<float, 2> yaw_cos = xt::cos(yaw_offseted);
-  xt::xtensor<float, 2> yaw_sin = xt::sin(yaw_offseted);
+  auto && yaw_cos = xt::xtensor<float, 1>::from_shape(traj_yaws.shape());
+  auto && yaw_sin = xt::xtensor<float, 1>::from_shape(traj_yaws.shape());
 
-  auto & vx = state.vx;
-  xt::xtensor<float, 2> dx = vx * yaw_cos;
-  xt::xtensor<float, 2> dy = vx * yaw_sin;
+  auto yaw_offseted = xt::view(traj_yaws, xt::range(1, _));
+
+  xt::noalias(xt::view(yaw_cos, 0)) = std::cos(initial_yaw);
+  xt::noalias(xt::view(yaw_sin, 0)) = std::sin(initial_yaw);
+  xt::noalias(xt::view(yaw_cos, xt::range(1, _))) = xt::cos(yaw_offseted);
+  xt::noalias(xt::view(yaw_sin, xt::range(1, _))) = xt::sin(yaw_offseted);
+
+  auto && dx = xt::eval(vx * yaw_cos);
+  auto && dy = xt::eval(vx * yaw_sin);
 
   if (isHolonomic()) {
-    auto & vy = state.vy;
     dx = dx - vy * yaw_sin;
     dy = dy + vy * yaw_cos;
   }
 
-  x = state_.pose.pose.position.x + xt::cumsum(dx * settings_.model_dt, 1);
-  y = state_.pose.pose.position.y + xt::cumsum(dy * settings_.model_dt, 1);
+  xt::noalias(traj_x) = state_.pose.pose.position.x + xt::cumsum(dx * settings_.model_dt, 0);
```

Comment:
Is there a reason why these are `0`s in this implementation but `1`s in the other?

Reply:
yeah, here we have just one trajectory as an input, so the dimension by which we apply cumsum becomes 0